### PR TITLE
Nightshift improvements: public area settings/levels, nightshift auth requirement change, config changes, etc etc

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request

Nightshift is much more configurable. 

## Why It's Good For The Game

People hate nightshifts as they are now. This makes them less obnoxious by default, and able to be by default set without an ID in most places.
Having the functionality is nice for a light dim switch.

## Changelog
:cl:
config: Added a few more nightshift config entries
add: nightshift_public_area variable in areas to determine how public an area is.
add: NIGHT_SHIFT_PUBLIC_AREAS_ONLY config entry allows the server to be configured to only nightshift areas of that level and below (so areas that are more public)
add: Config entries added for requiring authorizations to toggle nightshift. Defaults to only requiring on public areas as determined by above
/:cl:

### CODE REVIEW

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -959,70 +962,73 @@
...
 			update_icon()
 			update()
-		if("emergency_lighting")
-			emergency_lights = !emergency_lights
-			for(var/obj/machinery/light/L in area)
-				if(!initial(L.no_emergency)) //If there was an override set on creation, keep that override
-					L.no_emergency = emergency_lights
-					INVOKE_ASYNC(L, /obj/machinery/light/.proc/update, FALSE)
-				CHECK_TICK
+	var/authorized = (!locked || !usr.has_unlimited_silicon_privilege || failure_timer || (integration_cog && (is_servant_of_ratvar(usr))))
```
Comment (Ghommie) [inline code/modules/power/apc.dm]:
```suggestion
	var/authorized = (!locked || usr.has_unlimited_silicon_privilege || (integration_cog && (is_servant_of_ratvar(usr))))
```

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -959,70 +962,73 @@
...
+	if((action == "toggle_nightshift") && (!nightshift_requires_auth || authorized))
+		toggle_nightshift_lights()
+		. = TRUE
+	if(authorized)
```
Comment (Ghommie) [inline code/modules/power/apc.dm]: Early return.

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -959,70 +962,73 @@
...
-			emergency_lights = !emergency_lights
-			for(var/obj/machinery/light/L in area)
-				if(!initial(L.no_emergency)) //If there was an override set on creation, keep that override
-					L.no_emergency = emergency_lights
-					INVOKE_ASYNC(L, /obj/machinery/light/.proc/update, FALSE)
-				CHECK_TICK
+	var/authorized = (!locked || !usr.has_unlimited_silicon_privilege || failure_timer || (integration_cog && (is_servant_of_ratvar(usr))))
+	if((action == "toggle_nightshift") && (!nightshift_requires_auth || authorized))
+		toggle_nightshift_lights()
+		. = TRUE
```
Comment (Ghommie) [inline code/modules/power/apc.dm]:
```suggestion
		return TRUE
```

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by dapnee** [comment]
Kevin broke APCs, AI's can't interact with them thanks.

**Review by Ghommie — APPROVED** [review]

**Comment by Ghommie** [comment]
The interaction issue was pointed out and fixed.

**Pull request merged**